File: ocaml/day2/day2.ml
module Hand = struct
  type t = Rock | Paper | Scissor

  let create = function
      "A" | "X" -> Rock
      | "B" | "Y" -> Paper
      | "C" | "Z" -> Scissor
      | _ -> raise (Failure "not a valid hand")

  let get_value = function
      Rock -> 1
    | Paper -> 2
    | Scissor -> 3

  let winning = function
      Rock -> Paper
    | Paper -> Scissor
    | Scissor -> Rock

  let losing h = winning @@ winning h

  let cmp hand1 hand2 =
    match (hand1, hand2) with
      (Rock, Scissor) -> -1
    | (Scissor, Rock) -> 1
    | _ -> Int.compare (get_value hand2) (get_value hand1)
end

module Result = struct
  type t = Win | Draw | Lose

  let create = function
      "X" -> Lose
    | "Y" -> Draw
    | "Z" -> Win
    | _ -> raise (Failure "not a valid result")

  let get_value = function
      Win -> 6
    | Draw -> 3
    | Lose -> 0
end

let list_to_tuple = function
    a :: b :: [] -> (a, b)
  | _ -> raise (Failure "not a tuple")

let input =
  In_channel.with_open_bin "day2/input" In_channel.input_all
  |> String.split_on_char '\n'
  |> List.filter (fun s -> (String.length s != 0))
  |> List.map (fun s -> list_to_tuple @@ String.split_on_char ' ' s)

let first_strategy hands =
  let (elf_hand, my_hand) = hands in
  let my_hand_value = Hand.get_value my_hand in
  match Hand.cmp elf_hand my_hand with
    1 -> my_hand_value + Result.get_value Win
  | -1 -> my_hand_value + Result.get_value Lose
  | _ -> my_hand_value + Result.get_value Draw

let part1 =
  input
  |> List.map (fun (a, b) -> Hand.create a, Hand.create b)
  |> List.map first_strategy
  |> List.fold_left (+) 0

let second_strategy (elf_hand, (wanted_result: Result.t)) =
  match wanted_result with
    Win -> first_strategy (elf_hand, Hand.winning elf_hand)
  | Lose -> first_strategy (elf_hand, Hand.losing elf_hand)
  | Draw -> first_strategy(elf_hand, elf_hand)

let part2 =
  input
  |> List.map (fun (a, b) -> Hand.create a, Result.create b)
  |> List.map second_strategy
  |> List.fold_left (+) 0

let () =
  Printf.printf "Part 1 result: %d\n" part1;
  Printf.printf "Part 2 result: %d\n" part2;

File: rust/src/day2.rs
use std::{str::FromStr, cmp::Ordering};

use anyhow::Result;

#[derive(PartialEq)]
enum Hand {
    Rock = 1,
    Paper = 2,
    Scissor = 3,
}

impl FromStr for Hand {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "A" | "X" => Ok(Hand::Rock),
            "B" | "Y" => Ok(Hand::Paper),
            "C" | "Z" => Ok(Hand::Scissor),
            _ => Err(String::from("cannot parse this string to a hand"))
        }
    }
}

impl PartialOrd for Hand {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self == &Hand::Rock && other == &Hand::Scissor {
            Some(Ordering::Greater)
        } else if self == &Hand::Scissor && other == &Hand::Rock {
            Some(Ordering::Less)
        } else {
            Some((*self as u8).cmp(&(*other as u8)))
        }
    }
}

enum Outcome {
    Win,
    Draw,
    Lose,
}

impl FromStr for Outcome {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "X" => Ok(Outcome::Lose),
            "Y" => Ok(Outcome::Draw),
            "Z" => Ok(Outcome::Win),
            _ => Err(String::from("cannot parse this string to an outcome")),
        }
    }
}

fn get_result(hand1: &Hand, hand2: &Hand) -> u32 {
    match hand1.partial_cmp(hand2).unwrap() {
        Ordering::Less => 6 + *hand2 as u32, 
        Ordering::Greater => 0 + *hand2 as u32,
        Ordering::Equal => 3 + *hand2 as u32,
    }
}

fn part1() -> Result<u32> {
    Ok(
        aoc::read_file_as_lines::<String>("day2")?
            .into_iter()
            .map(|line| {
                let hands: Vec<Hand> = line
                    .split(" ")
                    .map(|s| s.parse::<Hand>().unwrap())
                    .collect();

                get_result(&hands[0], &hands[1])
            })
            .sum()
    )
}

fn part2() -> Result<u32> {
    Ok(
        aoc::read_file_as_lines::<String>("day2")?
            .into_iter()
            .map(|line| {
                let strs: Vec<&str> = line.split(" ").collect();

                let elf_hand = strs[0].parse::<Hand>().unwrap();
                let outcome = strs[1].parse::<Outcome>().unwrap();

                match outcome {
                    Outcome::Win => {
                        let my_hand = match elf_hand {
                            Hand::Rock => Hand::Paper,
                            Hand::Paper => Hand::Scissor,
                            Hand::Scissor => Hand::Rock,
                        };
                        6 + my_hand as u32
                    },
                    Outcome::Draw => 3 + elf_hand as u32,
                    Outcome::Lose => {
                        let my_hand = match elf_hand {
                            Hand::Rock => Hand::Scissor,
                            Hand::Paper => Hand::Rock,
                            Hand::Scissor => Hand::Paper,
                        };
                        0 + my_hand as u32
                    },
                }
            })
            .sum()
    )
}

fn main() -> Result<()> {
    println!("Part 1 result: {}", part1()?);
    println!("Part 2 result: {}", part2()?);

    Ok(())
}
